text/0000-guaranteed-tco.md

```diff
@@ -0,0 +1,602 @@
+<!-- This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work. -->
+This explanation is mostly based on a [previous RFC](https://github.com/DemiMarie/rfcs/blob/become/0000-proper-tail-calls.md#detailed-design)
+though is more restricted as the current RFC does not target general tail calls anymore.
+
+The goal of this RFC is to describe a first implementation that is already useful while providing a basis to explore
+possible ways to relax the requirements when TCO can be guaranteed.
+
+## Syntax
+[syntax]: #syntax
+
+A guaranteed TCO is indicated by using the `become` keyword in place of `return`. The `become` keyword is already
+reserved, so there is no backwards-compatibility break. The `become` keyword must be followed by a plain function call
+or method calls, that is supported are calls like: `become foo()`, `become foo(a)`, `become foo(a, b)`, and so on, or
+`become foo.bar()` with plain arguments. Neither the function call nor any arguments can be part of a larger expression
+such as `become foo() + 1`, `become foo(1 + 1)`, `become foo(bar())`. Additionally, there is a further restriction on
+the tail-callable functions: the function signature must exactly match that of the calling function. 
+
+Invocations of overloaded operators with at least one non-primitive argument were considered as valid targets, but were
+rejected on grounds of being too error-prone. In any case, these can still be called as methods.
+
+## Type checking
+[typechecking]: #typechecking
+A `become` statement is type-checked like a `return` statement, with the added restriction that the function signatures of the caller and callee must match exactly. Additionally, the caller and callee **must** use the same calling
+convention.
+
+## Borrowchecking and Runtime Semantics
+[semantics]: #semantics
+A `become` expression acts as if the following events occurred in-order:
+
+1. All variables that are being passed by-value are moved to temporary storage.
+2. All local variables in the caller are destroyed according to usual Rust semantics. Destructors are called where
+   necessary. Note that values moved from in step 1 are _not_ dropped.
+3. The caller's stack frame is removed from the stack.
+4. Control is transferred to the callee's entry point.
+
+This implies that it is invalid for any references into the caller's stack frame to outlive the call. The borrow checker ensures that none of the above steps will result in the use of a value that has gone out of scope.
+
+As `become` is always in a tail position (due to being used in place of `return`), this requirement for TCO is already
+fulfilled.
+
+Example:
```

Comment:
It is intended, the [template](https://github.com/phi-go/rfcs/blob/guaranteed-tco/0000-template.md#reference-level-explanation) asks to return to the examples that were used earlier. Though I guess this can also be done by reference.

Reply:
updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/fd35d0f4428f8e7cde72fa78052572ae554de629